### ISSUE TITLE
Standardise on a reconcile.Result pattern

### ISSUE THREAD

**Issue by negz** [crossplaneio/crossplane#447]
We do not currently have any consistency in how we return a `reconcile.Result` from a Crossplane controller. This is related to #419 and #302.

Many controllers simply return `reconcile.Result{}` or `reconcile.Result{Requeue: true}`. Several establish package scoped convenience variables with names like `result`, `resultDone` `resultRequeue`, `requeueOnSuccess`, or`requeueOnWait`. We've also introduced `RequeueOnWait` and `RequeueOnSuccess` constants [under `pkg/controller/core`](https://github.com/crossplaneio/crossplane/blob/6f3bace/pkg/controller/core/resource_controller.go#L51) representing the durations controllers _might_ choose to wait before requeuing by returning `reconcile.Result{RequeueAfter: core.RequeueOnSuccess}`.

Our tentative reconciler patterns design document [dedicates several paragraphs](https://github.com/crossplaneio/crossplane/blob/fbd063e/design/reconciler-patterns.md#output) to the intricacies and nuances of determining how a controller should requeue under what circumstances, indicating that choosing how to requeue is not obvious or intuitive for controller authors. I've made a handful of attempts to land on a set of clearly named and documented convenience variables in https://github.com/crossplaneio/crossplane/pull/424, but have not been able to convince other maintainers of their merit.

**Comment by negz** [comment]
I believe we _do_ have some consensus on the fact that it's _in most_ (but certainly not all) cases better to use `RequeueAfter` than `Requeue: true`. We also agree that there's no one "correct" amount of time to wait before requeuing - each controller and situation's needs will be different.

I would suggest that we've also found that it's _really hard_ to settle on any meaningful self-documenting names for reconcile results given how nuanced requeue decisions are. I feel that "success" and "on wait" are at best unclear, and at worst misleading, and have not been able to come up with anything obviously better.

Given the above my suggestion would be that we rely on documentation and PR review to educate new contributors about the intricacies of requeue decisions and explicitly return either `reconcile.Result{Requeue: true}`, `reconcile.Result{Requeue: false}`, or `reconcile.Result{RequeueAfter: 2 * time.Minute}` and remove our existing convenience variables for results or durations. Note that I suggest writing `reconcile.Result{Requeue: false}` because I feel it communicates intent (a requeue decision) more clearly than `reconcile.Result{}`'s implicit zero value. Also note that `2 * time.Minute` is just an example and that I expect the duration to vary from situation to situation.

**Comment by negz** [comment]
I've been experimenting with some controller refactoring, and found myself using a variant of the method I suggested in https://github.com/crossplaneio/crossplane/issues/447#issuecomment-491187479. In this variant I do use some constants for `RequeueAfter` times, but otherwise explicitly write out my desired `reconcile.Result{}` struct and include a comment explaining my requeue decision any time I `return reconcile.Result{...}, err`.

I've found myself using one of the below results as appropriate:
* `reconcile.Result{Requeue: false}`, meaning that I would not want this reconcile to requeue unless one happens implicitly due to an update or error.
* `reconcile.Result{RequeueAfter: CoolDuration}`, meaning that I would want this reconcile to requeue after the specified duration unless one happens implicitly due to an update or error.
* `reconcile.Result{}`, meaning that I _know_ this reconcile will definitely be requeued implicitly due to an update or error, so I'm explicitly omitting a requeue decision. This is identical in practice to `reconcile.Result{Requeue: false}`, but I think the fact I'm not explicitly setting `Requeue: false` makes my intent a little clearer.

I think the key thing that makes this method feel nice are the comments explaining the subtleties of every single requeue decision. It's quite verbose, but given how surprising requeue semantics can be I think it could work well.

**Comment by negz** [comment]
https://github.com/kubernetes-sigs/controller-runtime/issues/417 - Looks like requeue semantics are likely to change in a future version of controller-runtime.

**Comment by negz** [comment]
Below are some examples of the approach I suggested in https://github.com/crossplaneio/crossplane/issues/447#issuecomment-498535710 in practice.

```go
// Controller scoped RequeueAfter durations.
const (
	AShortWait = 30 * time.Second
	ALongWait  = 2 * time.Minute
)

if err := r.kube.Get(ctx, req.NamespacedName, claim); err != nil {
	// There's no need to requeue because we no longer exist. Otherwise
	// we'll be requeued implicitly because we return an error.
	return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), "cannot get resource claim")
}

if WasDeleted(claim) {
	delete := NewDeleteFn(r.kube)
	if err := delete(ctx, claim, resource); err != nil {
		// If we didn't hit this error last time we'll be requeued
		// implicitly due to the status update. Otherwise we want to retry
		// after a brief wait, in case this was a transient error.
		claim.SetConditions(corev1alpha2.ReconcileError(err))
		return reconcile.Result{RequeueAfter: AShortWait}, errors.Wrap(r.kube.Status().Update(ctx, claim), "cannot update resource claim status")
	}

	// We've successfully processed the delete, so there's no further
	// reconciliation to do. There's a good chance our claim no longer
	// exists, but we try update its status just in case it sticks around,
	// for example due to additional finalizers.
	return reconcile.Result{Requeue: false}, errors.Wrap(IgnoreNotFound(r.kube.Status().Update(ctx, claim)), "cannot update resource claim status")
}
```

**Comment by jbw976** [comment]
Addressed in #584
